lib/gc/ExecutionEngine/CPURuntime/Memorypool.cpp

```diff
@@ -0,0 +1,215 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#ifdef _WIN32
+#include <Windows.h>
+#else
+#include <sys/mman.h>
+#include <unistd.h>
+#endif
+
+#ifdef _MSC_VER
+#define __builtin_expect(EXP_, C) (EXP_)
+#endif
+
+#define likely(x) __builtin_expect(!!(x), 1)
+#define unlikely(x) __builtin_expect(!!(x), 0)
+
+namespace mlir {
+namespace gc {
+
+namespace memory_pool {
+static constexpr size_t divide_and_ceil(size_t x, size_t y) {
+  return (x + y - 1) / y;
+}
+static constexpr size_t default_alignment = 64;
+
+size_t get_os_page_size() {
+#ifdef _WIN32
+  // fix-me: (win32) impl
+  return 4096;
+#else
+  static size_t v = getpagesize();
+  return v;
+#endif
+}
+
+memory_chunk_t *memory_chunk_t::init(intptr_t pdata, size_t sz) {
+  memory_chunk_t *ths =
+      reinterpret_cast<memory_chunk_t *>(pdata - sizeof(memory_chunk_t));
+  ths->canary_ = magic_check_num_;
+  ths->size_ = sz;
+  return ths;
+}
+
+intptr_t memory_block_t::calc_alloc_ptr() {
+  intptr_t start_addr =
+      reinterpret_cast<intptr_t>(this) + allocated_ + sizeof(memory_chunk_t);
+  return divide_and_ceil(start_addr, default_alignment) * default_alignment;
+}
+
+void *alloc_by_mmap(size_t sz) {
+#ifdef _MSC_VER
+  auto ret =
+      VirtualAlloc(nullptr, sz, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+#else
+  auto ret = mmap(nullptr, sz, PROT_READ | PROT_WRITE,
+                  MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+#endif
+  assert(ret);
+  return ret;
+}
+
+memory_block_t *memory_block_t::make(size_t sz, memory_block_t *prev,
+                                     memory_block_t *next) {
+  auto ret = alloc_by_mmap(sz);
+  if (!ret) {
+    std::cerr << "Out of Memory." << std::endl;
+  }
+  memory_block_t *blk = reinterpret_cast<memory_block_t *>(ret);
+  blk->size_ = sz;
+  blk->allocated_ = sizeof(memory_block_t);
+  static_assert(sizeof(memory_block_t) == offsetof(memory_block_t, buffer_),
+                "sizeof(memory_block_t) == offsetof(memory_block_t, buffer_)");
+  blk->prev_ = prev;
+  blk->next_ = next;
+  return blk;
+}
+
+void dealloc_by_mmap(void *b) {
+#ifdef _MSC_VER
+  auto ret = VirtualFree(b, 0, MEM_RELEASE);
+  SC_UNUSED(ret);
+  assert(ret);
+#else
+  munmap(b, reinterpret_cast<memory_block_t *>(b)->size_);
+#endif
+}
+
+static void free_memory_block_list(memory_block_t *b) {
+  while (b) {
+    memory_block_t *next = b->next_;
+    dealloc_by_mmap(b);
+    b = next;
+  }
+}
+
+size_t filo_memory_pool_t::get_block_size(size_t sz) const {
+  // calculate the aligned size of management blocks in the header
+  constexpr size_t header_size =
+      divide_and_ceil(sizeof(memory_block_t) + sizeof(memory_chunk_t),
+                      default_alignment) *
+      default_alignment;
+  // the allocated size should include the aligned header size
+  sz = sz + header_size;
+  if (sz > block_size_) {
+    return divide_and_ceil(sz, get_os_page_size()) * get_os_page_size();
+  } else {
+    return block_size_;
+  }
+}
+
+void *filo_memory_pool_t::alloc(size_t sz) {
```

Comment:
do we need to update the coding style? `FILOMemoryPool`?

Reply:
Align with the MLIR coding style.